assignment.py

```diff
@@ -0,0 +1,135 @@
+def load_data():
+    current_year = get_current_year()
+
+    courses = [serialize_model(course) for course in
+               db.session.query(Course).filter_by(year=current_year).order_by(Course.quadri).all() or []]
+    supervisors = [serialize_model(supervisor) for supervisor in db.session.query(ResearcherSupervisor).all() or []]
+    teachers = {teacher.id: serialize_model(teacher) for teacher in
+                db.session.query(Teacher).filter_by(course_year=current_year).all() or []}
+    researchers = {researcher.id: serialize_model(researcher) for researcher in
+                   db.session.query(Researcher).all() or []}
+    preferences = {preference.id: serialize_model(preference) for preference in
+                   db.session.query(PreferenceAssignment).filter_by(course_year=current_year).all() or []}
+    organizations = {organization.id: serialize_model(organization) for organization in
+                     db.session.query(Organization).all() or []}
+
+    saved_data = [serialize_model(assignment) for assignment in
+                  db.session.query(AssignmentDraft).filter_by(course_year=current_year).all()]
+
+    # Retrieve active users or users associated with preferences.
+    # Join the 'Researcher' model to 'PreferenceAssignment' and check if the researcher has preferences.
+    users = {
+        user.id: serialize_model(user)
+        for user in db.session.query(User)
+        .filter(
+            (User.active == True) | (User.id.in_(
+                db.session.query(Researcher.user_id).join(PreferenceAssignment,
+                                                          Researcher.id == PreferenceAssignment.researcher_id)
+            ))
+        ).all()
+    }
+
+    data = {
+        'courses': courses,
+        'users': users,
+        'supervisors': supervisors,
+        'teachers': teachers,
+        'researchers': researchers,
+        'preferences': preferences,
+        'organizations': organizations,
+        'current_year': current_year,
+        'saved_data': saved_data,
+        'MAX_LOAD': DEFAULT_MAX_LOAD,
+    }
+
+    return jsonify(data)
+
+
+@assignment_bp.route('/publish_assignments', methods=['POST'])
+@login_required
+def publish_assignments():
+    data = request.get_json()
+    if not data:
+        return jsonify({"error": "No data provided"}), 400
+
+    current_year = get_current_year()
+    is_draft = data.get('isDraft')
+
+    try:
+        # Clear existing assignments for the current year
+        AssignmentDraft.query.filter_by(course_year=current_year).delete()
+        if not is_draft:
+            AssignmentPublished.query.filter_by(course_year=current_year).delete()
+        db.session.commit()
+
+        assignments_to_add = []
+
+        for item in data.get('data', []):
+            try:
+                user_data = item.get('userData')
+                course_data = item.get('courseData')
+
+                if not course_data or not user_data:
```

Comment:
If there is no assignment made for a researcher (empty line), `course_data` is empty and a `ValueError` is thrown: not sure this is an expected behaviour.

Besides, there would still be handy to know where the error is. In practice there will be more than a hundred lines in this table.